### PR TITLE
searcher: enforce sub-repo perms at api level

### PR DESCRIPTION
**Pull request by bobheadxi**
A crude implementation of #26647 by implementing enforcement as high up as possible, right before matches get sent to the client. I believe this covers indexed/unindexed/etc search that goes over repository file contents.

This requires adding `searcher` to the allowlist of `dbconn` importers.

### CODE REVIEW

File: internal/authz/sub_repo_perms.go [hunk 1]
```diff
@@ -78,6 +74,11 @@ func (s *SubRepoPermsClient) Permissions(ctx context.Context, userID int32, cont
 		return Read, nil
 	}
 
+	// Reject anonymous users
+	if userID == 0 {
+		return None, &ErrUnauthenticated{}
+	}
```
Comment (bobheadxi) [inline internal/authz/sub_repo_perms.go:80]: I realize I could do this right after merging https://github.com/sourcegraph/sourcegraph/pull/27038 🤦‍♂️ see `Actor.UID` docstring

File: internal/authz/sub_repo_perms.go [hunk 2]
```diff
@@ -26,10 +26,16 @@ type RepoContent struct {
...
+	// the provided context has for the requested content.
 	//
 	// If the context is unauthenticated, ErrUnauthenticated is returned.
 	CurrentUserPermissions(ctx context.Context, content RepoContent) (Perms, error)
+
+	// Permissions returns the level of access the provided user has for the requested
+	// content.
+	//
+	// If the userID represents an anonymous user, ErrUnauthenticated is returned.
+	Permissions(ctx context.Context, userID int32, content RepoContent) (Perms, error)
```
Comment (bobheadxi) [inline internal/authz/sub_repo_perms.go:38]: @ryanslade I ended up exporting this in the interface since in this case, using `CurrentUserPermissions` would mean reflecting `Actor` for each file.

File: cmd/searcher/protocol/searcher.go [hunk 3]
```diff
@@ -58,6 +61,23 @@ type Request struct {
...
+	}
+	// Surprisingly this is the same sanity check used in the git source.
+	if len(p.Commit) != 40 {
+		return errors.Errorf("Commit must be resolved (Commit=%q)", p.Commit)
+	}
+	if p.Pattern == "" && p.ExcludePattern == "" && len(p.IncludePatterns) == 0 {
+		return errors.New("At least one of pattern and include/exclude pattners must be non-empty")
+	}
+	if p.IsNegated && p.IsStructuralPat {
+		return errors.New("Negated patterns are not supported for structural searches")
```
Comment (unknwon) [inline cmd/searcher/protocol/searcher.go:76]:
```suggestion
		return errors.New("repo must be non-empty")
	}
	// Surprisingly this is the same sanity check used in the git source.
	if len(p.Commit) != 40 {
		return errors.Errorf("commit must be resolved (Commit=%q)", p.Commit)
	}
	if p.Pattern == "" && p.ExcludePattern == "" && len(p.IncludePatterns) == 0 {
		return errors.New("at least one of pattern and include/exclude pattners must be non-empty")
	}
	if p.IsNegated && p.IsStructuralPat {
		return errors.New("negated patterns are not supported for structural searches")
```

nit: error strings start with lowercase as the general Go convention :D

**Comment by stefanhengl** [comment]
I don't think indexed search is covered by this. In the description you mention "indexed search" and #26647 mentions Zoekt but it seems the scope was changed to searcher?

**Comment by ryanslade** [comment]
@bobheadxi Oops, I didn't see this PR until after I merged https://github.com/sourcegraph/sourcegraph/pull/27054

Sorry for the conflict

File: cmd/searcher/search/search.go [hunk 4]
```diff
@@ -100,12 +102,32 @@ func (s *Service) streamSearch(ctx context.Context, w http.ResponseWriter, p pro
...
+
 	matchesBuf := streamhttp.NewJSONArrayBuf(32*1024, func(data []byte) error {
 		return eventWriter.EventBytes("matches", data)
 	})
 	onMatches := func(match protocol.FileMatch) {
+		// Drop all results that do user has insufficient permissions for.
+		//
+		// 🚨 SECURITY: This must be correctly implemented to hide anything about
+		// files that a user should not see.
+		perms, err := s.SubRepoPerms.Permissions(ctx, a.UID, authz.RepoContent{
```
Comment (keegancsmith) [inline cmd/searcher/search/search.go:115]: how fast is this? does it always speak to the DB? Is it fast enough to run before searching a file? If you are filtering here, it almost feels like you might as well filter in the searcher client and avoid the DB dep.

File: cmd/searcher/search/search.go [hunk 5]
```diff
@@ -100,12 +102,32 @@ func (s *Service) streamSearch(ctx context.Context, w http.ResponseWriter, p pro
...
+			Repo: p.Repo,
+			Path: match.Path,
+		})
+		if err != nil {
+			s.Log.Warn("streamSearch.onMatches: error when validating sub-repo permissions",
+				"error", err,
+				"repo", p.Repo)
```
Comment (keegancsmith) [inline cmd/searcher/search/search.go:122]: nit I know you check again that err != nil and then return. But it feels more correct to just return after this statement and skip the extra err check below.

File: cmd/searcher/search/search.go [hunk 6]
```diff
@@ -100,12 +102,32 @@ func (s *Service) streamSearch(ctx context.Context, w http.ResponseWriter, p pro
...
+		// Drop all results that do user has insufficient permissions for.
+		//
+		// 🚨 SECURITY: This must be correctly implemented to hide anything about
+		// files that a user should not see.
+		perms, err := s.SubRepoPerms.Permissions(ctx, a.UID, authz.RepoContent{
+			Repo: p.Repo,
+			Path: match.Path,
+		})
+		if err != nil {
+			s.Log.Warn("streamSearch.onMatches: error when validating sub-repo permissions",
```
Comment (keegancsmith) [inline cmd/searcher/search/search.go:120]: should this be a warn? I imagine this client failing is a pretty bad sign?

File: cmd/searcher/protocol/searcher.go [hunk 7]
```diff
@@ -58,6 +61,23 @@ type Request struct {
 	Indexed bool
 }
 
+func (p *Request) ValidateParams() error {
```
Comment (keegancsmith) [inline cmd/searcher/protocol/searcher.go:64]: why was this moved and exported? I don't see any new uses of it, it seems reasonable to keep it as a private implementation detail of the server.

File: cmd/searcher/search/search.go [hunk 8]
```diff
@@ -100,12 +102,32 @@ func (s *Service) streamSearch(ctx context.Context, w http.ResponseWriter, p pro
 		return
 	}
 
+	a := actor.FromContext(ctx)
```
Comment (keegancsmith) [inline cmd/searcher/search/search.go:105]: have you confirmed we correctly transport the actor across our RPC layer? We have never needed it before, so I'd be surprised if this just works.

File: internal/authz/sub_repo_perms.go (hunk 1 above)
Comment (keegancsmith) [reply internal/authz/sub_repo_perms.go:80]: I think we may have internal users of this which will have internal users. So before shipping this I would confirm that UID is always set in searcher with some sort of metric?

File: cmd/searcher/search/search_test.go [hunk 9]
```diff
@@ -48,126 +51,127 @@ func main() {
 
 	cases := []struct {
 		arg  protocol.PatternInfo
+		srp  authz.SubRepoPermissionChecker
```
Comment (keegancsmith) [inline cmd/searcher/search/search_test.go:54]: nit: rather than specifying a permission checker per test case, why not just have a global one and have test cases which exercise it. Makes the test cases much more readable. Same idea above where the files are global. You also avoid starting up an HTTP server per test case as well, which seems like overkill.

**Review by keegancsmith — COMMENTED** [review]
Cool. From a high level what made you decide on this layer vs others? (eg in frontend, or lower down in searcher where we decide to look at a file). Is there more info on how the client works and the performance characteristics available / will be available.

File: cmd/searcher/search/search.go [hunk 10]
```diff
@@ -100,12 +102,32 @@ func (s *Service) streamSearch(ctx context.Context, w http.ResponseWriter, p pro
...
+
 	matchesBuf := streamhttp.NewJSONArrayBuf(32*1024, func(data []byte) error {
 		return eventWriter.EventBytes("matches", data)
 	})
 	onMatches := func(match protocol.FileMatch) {
+		// Drop all results that do user has insufficient permissions for.
```
Comment (ryanslade) [inline cmd/searcher/search/search.go:111]:
```suggestion
		// Drop all results that the user has insufficient permissions for.
```

File: cmd/searcher/main.go [hunk 11]
```diff
@@ -67,6 +75,9 @@ func main() {
 			MaxCacheSizeBytes: cacheSizeBytes,
 		},
 		Log: log15.Root(),
+		SubRepoPerms: &authz.SubRepoPermsClient{
+			PermissionsGetter: database.SubRepoPerms(dbconn.Global),
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply cmd/searcher/main.go:79]: Should we just provide a constructor? I assumed that might be an optional optimization, but I suppose we should just have everything available up-front

[the remark this answers is not in the document]
Comment (bobheadxi) [reply cmd/searcher/main.go:79]: When the default is constructing by structs IMO all fields should be optional

File: cmd/searcher/main.go [hunk 12]
```diff
@@ -57,6 +60,11 @@ func main() {
 		cacheSizeBytes = i * 1000 * 1000
 	}
 
+	opts := dbconn.Opts{DSN: "", DBName: "searcher", AppName: "searcher"}
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply cmd/searcher/main.go:63]: The "smart client" is the `SubRepoPermsClient`, but I think you're right - I am going to have another go at this PR attempting filtering higher up. Had a bit of misunderstanding myself about the approach to "client side" - I was thinking this might be the point at which we go client -> zoekt/etc, but if most things are in searcher we could also go client -> searcher

File: cmd/searcher/search/search.go (hunk 4 above)
Comment (bobheadxi) [reply cmd/searcher/search/search.go:115]: You're right! I'm going to have another go doing this at the searcher client

File: cmd/searcher/protocol/searcher.go (hunk 7 above)
Comment (bobheadxi) [reply cmd/searcher/protocol/searcher.go:64]: It was otherwise difficult for me to tell what were required params (I was in particular not sure if `Repo` was required), though I'l leave it as is in my follow-up PR

**Comment by bobheadxi** [comment]
> I don't think indexed search is covered by this. In the description you mention "indexed search" and #26647 mentions Zoekt but it seems the scope was changed to searcher?

@stefanhengl ah you might be right - as I was looking at Zoekt client I think I traced it back up to searcher, I probably made a mistake here and conflated the two. Looking at `searcher/client.go` it does appear `Search` is only for unindexed search

**Comment by bobheadxi** [comment]
@keegancsmith 

> From a high level what made you decide on this layer vs others? (eg in frontend, or lower down in searcher where we decide to look at a file)

- Against higher up:
  - Concerns about missing usages of search APIs internally (though potentially not a concern if everything ends up going through frontend APIs anyway)
  - Potentially too far away from where results are actually being searched to implement pre-request optimizations that we want to make possible
- Against lower down: 
  - The search code seems to branch based on a variety of conditions, too far down and we'll have to implement in multiple places
  - I'm thinking we likely have to provide `SubRepoPermsClient` or similar down the call stack, due to database dependency. We can use a global database handle, but there's efforts to move away from that so I'd like to avoid adding to the issue. Too far down feels like a potentially too invasive change, and we want to get this implemented with minimal disruption where possible

**Comment by bobheadxi** [comment]
> Is there more info on how the client works and the performance characteristics available / will be available.

- How it works: internally, it will take our matching syntax and match the path against them. more docs will be available on this soon I think, but the concrete implementation of the client + https://github.com/sourcegraph/sourcegraph/pull/26745 are a good place to start looking for now
- Performance: We're focusing on making this functional first with minimal regards to performance, while ensuring a "fast path" is taken by default (so that all existing Sourcegraph instances are unaffected). We have some improvements in mind, e.g. caching #26663 , but that will likely come later

File: cmd/searcher/search/search.go (hunk 8 above)
Comment (bobheadxi) [reply cmd/searcher/search/search.go:105]: cc @ryanslade , might be relevant for #26646

**Comment by bobheadxi** [comment]
I've started [a thread in #search](https://sourcegraph.slack.com/archives/CHEKCRWKV/p1636056611076000) with some questions 🙏
